pkg/finder/crifinder.go

```diff
@@ -0,0 +1,334 @@
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	runtimespec "github.com/opencontainers/runtime-spec/specs-go"
+	"google.golang.org/grpc"
+
+	"k8s.io/api/core/v1"
+
+	pb "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+	"k8s.io/kubernetes/pkg/kubelet/util"
+)
+
+const (
+	defaultTimeout     = 5 * time.Second
+	ns                 = "resource-topology-exporter"
+	CGroupCPUSetPrefix = "fs/cgroup/cpuset"
+	CGroupCPUsetSuffix = "cpuset.cpus"
+	ContainerdRuntime  = "containerd"
+	CRIORuntime        = "cri-o"
+)
+
+type CGroupPathTranslator func(sysfs, cgroupPath string) string
+type criFinder struct {
+	args             Args
+	conn             *grpc.ClientConn
+	client           pb.RuntimeServiceClient
+	cgroupTranslate  CGroupPathTranslator
+	NodeResourceData *NodeResources
+}
+
+type ContainerInfo struct {
+	sandboxID      string              `json:"sandboxID"`
+	Pid            uint32              `json:"pid"`
+	Removing       bool                `json:"removing"`
+	SnapshotKey    string              `json:"snapshotKey"`
+	Snapshotter    string              `json:"snapshotter"`
+	RuntimeType    string              `json:"runtimeType"`
+	RuntimeOptions interface{}         `json:"runtimeOptions"`
+	Config         *pb.ContainerConfig `json:"config"`
+	RuntimeSpec    *runtimespec.Spec   `json:"runtimeSpec"`
+}
+
+func NewCRIFinder(args Args, pciResMapConf map[string]string) (Finder, error) {
+	finderInstance := &criFinder{
+		args: args,
+	}
+	var err error
+
+	// At this stage, we only support containerd and cri-o
+	if args.ContainerRuntime == ContainerdRuntime {
+		finderInstance.cgroupTranslate = containerDCGroupPathTranslate
+	} else {
+		//cri-o
+		finderInstance.cgroupTranslate = crioCGroupPathTranslate
+	}
+	finderInstance.NodeResourceData, err = NewNodeResources(finderInstance.args.SysfsRoot, pciResMapConf)
+	if err != nil {
+		return nil, err
+	}
+
+	// now we can connext to CRI
+	addr, dialer, err := getAddressAndDialer(finderInstance.args.CRISocketPath)
+	if err != nil {
+		return nil, err
+	}
+
+	finderInstance.conn, err = grpc.Dial(addr, grpc.WithInsecure(), grpc.WithBlock(), grpc.WithTimeout(defaultTimeout), grpc.WithContextDialer(dialer))
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect: %v", err)
+	}
+
+	finderInstance.client = pb.NewRuntimeServiceClient(finderInstance.conn)
+	log.Printf("connected to '%v'!", finderInstance.args.CRISocketPath)
+	if finderInstance.args.Namespace != "" {
+		log.Printf("watching namespace %q", finderInstance.args.Namespace)
+	} else {
+		log.Printf("watching all namespaces")
+	}
+
+	return finderInstance, nil
+}
+
+func getAddressAndDialer(endpoint string) (string, func(ctx context.Context, addr string) (net.Conn, error), error) {
+	return util.GetAddressAndDialer(endpoint)
+}
+
+func (f *criFinder) GetNodeResourceData() *NodeResources {
+	return f.NodeResourceData
```

Comment:
I think `Scan()` should return this data; IOW I don't see the value of this extra getter.

Reply:
Scan returns PodResourceData which is information about all the allocated resources to the pods whereas this getter returns nodeResourceData (pointer to instance NodeResources):
```
type NodeResources struct {
	pciDevs         *pcidev.PCIDevices
	cpus            *cpus.CPUs
	cpuID2NUMAID    map[int]int
	pciAddr2NUMAID  map[string]int
	perNUMACapacity map[int]map[v1.ResourceName]int64
	// pciaddr -> resourcename
	pci2ResourceMap map[string]string
}
```

nodeResourceData stores information of the resources available (CPUs obtained after scanning sysfs and devices obtained by using sr-iov config file) and  on the node on which the daemon is running. But, since we are assuming that the resources on the nodes will remain the same during the process life cycle, we can do this get this information once in main rather than having it in both criFinder and podResourceFinder structs. So, moving code pertaining to to NodeResources in a separate file and instantiating NodeResourceData in main now.